sync vhost with cloud-hypervisor/vhost_rs

eryugey opened the pull request:
This pull request syncs all modifications to cloud-hypervisor/vhost_rs to this vhost repo, with some improvements from Alibaba on top of it.

The purpose of this sync is to build a common code base that could be used by both cloud-hypervisor and Alibaba's dragonball, and maybe by other projects as well. So people could collaborate on the common library, and all could benefit from it.

This pull request is targeted at dragonball branch, I hope it could be merged into master branch when everything settles down, and cloud-hypervisor itself could switch to this crate soon, instead of maintaining a private cloud-hypervisor/vhost_rs code base.

I went through all commits that touched cloud-hypervisor/vhost_rs directory and ported them to vhost repo dragonball branch. The general rules are

- Perfer code from cloud-hypervisor/vhost_rs. So you may see some reverts, because cloud-hypervisor/vhost_rs has similar commits and the original commits are reverted.
- Preserve the original authors and dates, but change "vhost_rs" prefix to "vhost-user"
- Some commits touched code outside cloud-hypervisor/vhost_rs, which got deleted in this port
- All Cargo.toml updates are merged into one commit (the first patch)
- Commits from Alibaba are on top of the tree (the last two commits)

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

yangzhon committed:
vhost-user: The vhost user version we support

The vhost user version should be same with backend.
define VHOST_USER_VERSION    (0x1)

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
vhost-user: Allow MasterReqHandler to reply when needed

No matter if the communication is coming from the master or the slave,
it should always reply with an ack if the message header specifies that
it expects a reply.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

Commit:
vhost-user: Wait for full request to be satisfied

The recvmsg syscall can split a request in multiple packets unless we
use the flag MSG_WAITALL to make sure the request will wait for the
whole data to be transferred before returning.

This flag is needed to prevent the vhost crate from returning the error
PartialMessage, which occured sporadically when using virtio-fs, and
which was detected as part of our continuous integration testing.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

Commit:
vhost-user: Fix GET_CONFIG command

Following the vhost-user specification that can be found at
https://github.com/qemu/qemu/blob/master/docs/interop/vhost-user.rst#virtio-device-config-space
this patch fixes the GET_CONFIG command.

This command is very specific as it needs to provide a body and a
payload corresponding to the content of the configuration structure as
an array of bytes. As a reply, it receives the body and payload provided
by the slave.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

yangzhon committed:
vhost-user: Add INFLIGHT_SHMFD protocol feature

Add to the vhost_rs crate the INFLIGHT_SHMFD protocol feature that was
missing from the list. The feature has been recently introduced in
vhost-user specification, which is the reason why it was missing.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

rbradford committed:
vhost-user: Add clippy override

clippy wants us to use "if let Some(fd) = fd {}" but to combine that
with && is an experimental feature.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

slp committed:
vhost-user: fix VhostUserConfig payload management

The VhostUserConfig carries a message with a payload, the contents of
which depend on the kind of device being emulated.

With this change, we calculate the offset of the payload within the
message, check its size corresponds to the expected one, and pass it
to the backend as a reference to a slice adjusted to the payload
dimensions.

The backend will be responsible of validating the payload, as it's the
one aware of its expected contents.

Signed-off-by: Sergio Lopez <slp@redhat.com>

rbradford committed:
vhost-user: Mark some broken tests ignored

If we use the workspace based testing methodology then we start testing
some code inside vhost_rs that is broken.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vhost-user: Retry connections to vhost-user backends

If the connection to the backend fails, sleep and retry up to a count.
This failure could happen because the backend is not yet ready to handle
the connection which is a common case when working with a spawned
backend.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

Commit:
vhost-user: add SlaveFsCacheReq to handle map/unmap

This introduces SlaveFsCacheReq which implements
VhostUserMasterReqHandler to handle map/unmap requests.

Signed-off-by: Liu Bo <bo.liu@linux.alibaba.com>

Commit:
vhost-user: add the ability to set slave req fd

This adds the missing part of supporting virtiofs dax on the slave end,
that is, receiving a socket pair fd from the master end to set up a
communication channel for sending setupmapping & removemapping messages.

Signed-off-by: Liu Bo <bo.liu@linux.alibaba.com>

eryugey committed:
vhost-user: make VHOST_USER_FS_SLAVE_ENTRIES public

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

Commit:
vhost-user: Fix unit test race condition

The unit tests are run from cargo test through multiple threads of the
same process. For this reason, all these threads share their file
descriptors (because that's how this works on Linux), which means that
any of them can close a file descriptor opened from another thread.

In the context of create_listener() and accept_connection() tests, they
can run concurrently and this generates some failure when the file
descriptor create_listener() is binding to is being closed from the
accept_connection() test.

In order to avoid such race condition, this patch simply removes the
part of the unit test performing an explicit and unsafe file descriptor
closure.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

eryugey committed:
vhost-user: handle short read/write in vhost-user connections

The underlying sendmsg/recvmsg syscall could be interrupted by signals
and returns bytes less than asked, then methods like
send_header()/recv_header() would return Error::PartialMessage to
callers. This may cause VMM fail to start due to partial read/write.

Fix by retrying send/recv until all data has been transfered. This also
follows what qemu does in qemu's vhost_user code.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

eryugey committed:
vhost-user: retry on SocketRetry error in vhost-user connection

sendmsg/recvmsg syscalls could be interrupted by signals before any
data has been transfered, we should retry in this case.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

rbradford committed:
vhost-user: Return error when connection broken

Return an error wen recvmsg() returns without a message using the
libc::ECONNRESET error so that the upper levels will correctly
interpret this as the connection being broken.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

slp committed:
vhost-user: Don't check for SLAVE_SEND_FD on SET_SLAVE_REQ_FD

SLAVE_SEND_FD and SLAVE_REQ are different protocol features. Check for
SLAVE_REQ instead.

Signed-off-by: Sergio Lopez <slp@redhat.com>

eryugey committed:
Cargo: sync dependencies with cloud-hypervisor/vhost_rs

Update dependencies to sync with cloud-hypervisor/vhost_rs.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

eryugey committed:
Revert "master: enable config messge for get_config()"

This reverts commit 8aedf7685225909253a5887228107ff0ed5d68ae.

The GET_CONFIG command is fixed by commit 927148dd3cfc ("vhost_rs: Fix
GET_CONFIG command") from cloud-hypervisor in a similar way. Let's take
that commit and be synced with cloud-hypervisor.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

eryugey committed:
Revert "slave: Fix VhostUserConfig (un)marshalling"

This reverts commit 0ec363b01d189a46db7689c046627d33c46c28d4.

The VhostUserConfig payload is fixed by commit 85e936d4bde6 ("vhost_rs:
fix VhostUserConfig payload management") from cloud-hypervisor, let's
be synced with cloud-hypervisor.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

eryugey committed:
vhost-user: control SlaveFsCacheReq with vhost-user-slave feature

We import slave_fs_cache mod under vhost-user-slave feature control, but
not the self::slave_fs_cache::SlaveFsCacheReq import.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

sboeuf commented on src/vhost_user/master.rs:
```diff
@@ -356,12 +356,11 @@ impl VhostUserMaster for Master {
         }
 
-        // vhost-user spec states that:
+        // TODO: vhost-user spec states that:
```
Please remove the `TODO`.

eryugey replied:
I think this addition of "TODO" is due to a revert commit, because I want to keep a clean port:

commit a94720fc4b5d198db684c9e27cc7f221a23583bd
Author: Eryu Guan <eguan@linux.alibaba.com>
Date:   Mon Feb 24 16:04:11 2020 +0800

    Revert "master: enable config messge for get_config()"
    
    This reverts commit 8aedf7685225909253a5887228107ff0ed5d68ae.
    
    The GET_CONFIG command is fixed by commit 927148dd3cfc ("vhost_rs: Fix
    GET_CONFIG command") from cloud-hypervisor in a similar way. Let's take
    that commit and be synced with cloud-hypervisor.

And the TODO got deleted by a later patch:

commit 29fe7579670370f04a656c4f623141bf620b16a5
Author: Sebastien Boeuf <sebastien.boeuf@intel.com>
Date:   Tue Sep 10 15:29:53 2019 -0700

    vhost-user: Fix GET_CONFIG command
    
    Following the vhost-user specification that can be found at
    https://github.com/qemu/qemu/blob/master/docs/interop/vhost-user.rst#virtio-device-config-space
    this patch fixes the GET_CONFIG command.
    
    This command is very specific as it needs to provide a body and a
    payload corresponding to the content of the configuration structure as
    an array of bytes. As a reply, it receives the body and payload provided
    by the slave.
    
    Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf reviewed (approved):
Looks good!
The way you organized the PR makes it easy to review, and the changes you brought from Alibaba's side look fine to me :)

rbradford commented:
Hi, thanks for doing this.

a.) Can you create a PR for Cloud Hypervisor that makes it point at this repo/branch so we can be sure everything works well
b.) What's the path to getting this into rust-vmm/vhost ?

eryugey commented:
> Hi, thanks for doing this.
> 
> a.) Can you create a PR for Cloud Hypervisor that makes it point at this repo/branch so we can be sure everything works well

Sure, will do.

> b.) What's the path to getting this into rust-vmm/vhost ?

@jiangliu talked to @sameo a few days ago (offline), and we came to an agreement that we collaborate on cloud-hypervisor/vhost first, becasue AWS seems don't care much about their vhost crate, at least the features we want.

@jiangliu please comment if you have more info to provide.

jiangliu commented:
> > Hi, thanks for doing this.
> > a.) Can you create a PR for Cloud Hypervisor that makes it point at this repo/branch so we can be sure everything works well
> 
> Sure, will do.
> 
> > b.) What's the path to getting this into rust-vmm/vhost ?
> 
> @jiangliu talked to @sameo a few days ago (offline), and we came to an agreement that we collaborate on cloud-hypervisor/vhost first, becasue AWS seems don't care much about their vhost crate, at least the features we want.
> 
> @jiangliu please comment if you have more info to provide.

Yes, it would be great for CLH/Dragonball to adopt the vhost crate and make is as mature as possible. Once we have confidence that the vhost crate is ready for rust-vmm, we will issue PR again.

eryugey commented:
> Hi, thanks for doing this.
> 
> a.) Can you create a PR for Cloud Hypervisor that makes it point at this repo/branch so we can be sure everything works well

Done, please see https://github.com/cloud-hypervisor/cloud-hypervisor/pull/821

sameo commented:
@eryugey Could you please rebase on top of the dragonball branch? I added a gh action based CI and it would be good to verify everything passes.

eryugey commented:
> @eryugey Could you please rebase on top of the dragonball branch? I added a gh action based CI and it would be good to verify everything passes.

Sure, and done!

sameo commented:
@eryugey I think `cargo build --release --features=vhost-user-slave --target x86_64-unknown-linux-gnu` fails to build.

yangzhon commented:
> > @jiangliu talked to @sameo a few days ago (offline), and we came to an agreement that we collaborate on cloud-hypervisor/vhost first, becasue AWS seems don't care much about their vhost crate, at least the features we want.
> > @jiangliu please comment if you have more info to provide.
> 
> Yes, it would be great for CLH/Dragonball to adopt the vhost crate and make is as mature as possible. Once we have confidence that the vhost crate is ready for rust-vmm, we will issue PR again.

@jiangliu, vhost crate will be separated from virtio crate into rust-vmm? thanks.

eryugey commented:
> @eryugey I think `cargo build --release --features=vhost-user-slave --target x86_64-unknown-linux-gnu` fails to build.

Ah, that's because the newly added SlaveFsCacheReq code (src/vhost_user/slave_fs_cache.rs) depends on vhost-user-master feature as well, as it implements VhostUserMasterReqHandler trait.

The following diff fixes the issue for me, import master_req_handler if either vhost-user-master or vhost-user-slave feature is enabled.

```diff
diff --git a/src/vhost_user/mod.rs b/src/vhost_user/mod.rs
index 521c61ee8f71..88642b22febc 100644
--- a/src/vhost_user/mod.rs
+++ b/src/vhost_user/mod.rs
@@ -28,9 +28,9 @@ pub use self::connection::Listener;
 mod master;
 #[cfg(feature = "vhost-user-master")]
 pub use self::master::{Master, VhostUserMaster};
-#[cfg(feature = "vhost-user-master")]
+#[cfg(any(feature = "vhost-user-master", feature = "vhost-user-slave"))]
 mod master_req_handler;
-#[cfg(feature = "vhost-user-master")]
+#[cfg(any(feature = "vhost-user-master", feature = "vhost-user-slave"))]
 pub use self::master_req_handler::{MasterReqHandler, VhostUserMasterReqHandler};
 
 #[cfg(feature = "vhost-user-slave")]
```

This is a bug from the original code, I'll add a new patch to fix it.

But I did test build with "-all-features" option, not sure why I missed it..I usually do
```bash
cargo build --all-features --tests
```

eryugey committed:
vhost-user: SlaveFsCacheReq depends on vhost-user-slave feature as well

We implement VhostUserMasterReqHandler trait for SlaveFsCacheReq struct,
and VhostUserMasterReqHandler depends on vhost-user-master feature, so
cargo build with only vhost-user-slave feature set fails.

Fix it by importing VhostUserMasterReqHandler when either
vhost-user-master or vhost-user-slave feature is on.

Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

eryugey commented:
> But I did test build with "-all-features" option, not sure why I missed it..I usually do
> 
> ```shell
> cargo build --all-features --tests
> ```

--all-features means both vhost-user-master and vhost-user-slave are enabled, so I didn't hit this failure, that explains.. I'll improve my build test matrix.

eryugey commented:
New patch pushed:

f263f1fd30b1 vhost-user: SlaveFsCacheReq depends on vhost-user-slave feature as well

sameo commented:
@eryugey Nice, CI passes now, thanks.
One last thing before we can merge: There are two commits that do not carry a Signed-off-by line:

03699ee5b9b8db658d30250cfedafcb8b482dcf0
00c7cba20c2e3eb284053115b5a80c231dedd0a3

Would you mind fixing that as well? Sorry I did not notice earlier.

eryugey commented:
> @eryugey Nice, CI passes now, thanks.
> One last thing before we can merge: There are two commits that do not carry a Signed-off-by line:
> 
> [03699ee](https://github.com/cloud-hypervisor/vhost/commit/03699ee5b9b8db658d30250cfedafcb8b482dcf0)
> [00c7cba](https://github.com/cloud-hypervisor/vhost/commit/00c7cba20c2e3eb284053115b5a80c231dedd0a3)
> 
> Would you mind fixing that as well? Sorry I did not notice earlier.

They're the two revert commits made by me, and I forgot to add a SOB tag.. Sure, I'll fix.

Also, I noticed there's a new commit in cloud-hypervisor/vhost_rs, I'll add it as well.

d8d790bb7bf5 vhost_rs: Don't check for SLAVE_SEND_FD on SET_SLAVE_REQ_FD
